CUE Integration

bonedaddy opened the pull request:
# Purpose

Solves part of the requirement for #39, after this we can work on the Experiments API

# Notes

* I've made some assumptions about how the trials concept should be integrated into experiment definitions, and some of these assumptions are probably incorrect. 
* I found CUE notably difficult to work with, but this is likely due to me having never used configuration languages (or CUE) up until this PR beyond simple copy+paste from README's to run things
* There's a lot of JSON marshalling/unmarshalling, so it might be worth looking into [simdjson-go](https://github.com/minio/simdjson-go)

# Changes

* Enable `metadata.ExperimentDefinition` creation from CUE source files
  * This is done via a "parser" package
* Adds a `TrialDefinition` which is a combination of 1 or more Trials, with a trial being a combination of clusters + scenarios to run together in a benchmark
*`TrialDefinition` has been added as a field to the `metadata.ExperimentDefinition` struct

hinshun commented on cue/cue.mod/p2plab.cue:
```diff
@@ -0,0 +1,40 @@
+// defines a set of nodes of size 1 or higher
+// a "node" is simply an EC2 instance provisioned of the given type
+Nodes :: {
```
I think this should just be called `Group`.

bonedaddy replied:
Good idea, I did find `Nodes` a bit awkward.

hinshun commented on cue/cue.mod/p2plab.cue:
```diff
@@ -0,0 +1,40 @@
+    cluster: Cluster
+    scenario: Scenario
+    // trials: [ ... ]
+    trials: [ ...[Cluster,Scenario]]
```
Can this be just `[ ...Trial]`? Is it to reduce the verbosity of writing experiments by avoiding the `Trial` object wrapping?

bonedaddy replied:
I believe so, let me try

hinshun commented on cue/cue.mod/p2plab.cue:
```diff
@@ -0,0 +1,40 @@
+
+Experiment :: {
+    cluster: Cluster
+    scenario: Scenario
```
What's the purpose of these top-level cluster, scenario?

bonedaddy replied:
I thought this would be incorrect, but wasn't sure so I just left them in in case they were needed. Based off [this comment](https://github.com/Netflix/p2plab/pull/41#discussion_r413397157) I'll just scrap the top level cluster/scenario and just stick them in the trial type

hinshun commented on metadata/experiment.go:
```diff
@@ -50,6 +50,8 @@ type ExperimentDefinition struct {
 	ClusterDefinition ClusterDefinition
 
 	ScenarioDefinition ScenarioDefinition
```
Ah I see why you got confused before. We should remove these top-level cluster/scenario definitions.

bonedaddy replied:
Sounds good

hinshun commented on metadata/trial.go:
```diff
@@ -0,0 +1,12 @@
+
+// Trial is one particular cluster + scenario combination to run
+type Trial struct {
+	Cluster  []ClusterGroup     `json:"cluster"`
```
Use `ClusterDefinition` here instead of `[]ClusterGroup`.

bonedaddy replied:
sounds good

hinshun commented on metadata/experiment.go:
```diff
@@ -50,6 +50,8 @@ type ExperimentDefinition struct {
 
 	ScenarioDefinition ScenarioDefinition
+
+	TrialDefinition TrialDefinition
```
We should just have `Trials []TrialDefinition` here.

bonedaddy replied:
good idea that will make this a bit cleaner

bonedaddy commented:
thanks for the review, I will go about making these changes and let you know when they're done

bonedaddy commented:
Added the cue example you wrote as an additional source file to test, and removed the `ClusterDefinition` and `ScenarioDefinition` top level fields of `ExperimentDefinition`, let me know if there's anything else I should change/add.

Side note: This change makes cue a lot easier to use

hinshun commented:
Almost there, we still need to address this: https://github.com/Netflix/p2plab/pull/41/files#r413397302

bonedaddy commented:
> Almost there, we still need to address this: https://github.com/Netflix/p2plab/pull/41/files#r413397302

Ah yes, I forgot about this
